subprojects/core/src/main/java/org/gradle/api/internal/attributes/UsageCompatibilityHandler.java

```diff
@@ -34,6 +34,8 @@ class UsageCompatibilityHandler {
     }
 
     public <T> ImmutableAttributes doConcat(DefaultAttributesFactory factory, ImmutableAttributes node, Attribute<T> key, Isolatable<T> value) {
+        factory.assertAttributeNotAlreadyPresent(node, key);
```

Comment:
💭 Another potential route to strange failures - we should failfast if we try to add the same attribute twice when concat-ing.

Reply:
Does this trigger any failures in real builds? I don't think it should

Reply:
Ignore that last comment.  There is a behavior difference between `concat` and `sateConcat` that is required.  We make use of it when parsing the GMM.  The methods are misnamed, but functioning as desired - I'm adjusting the tests to account for this.